near-sdk/src/store/lookup_map/mod.rs

```diff
@@ -0,0 +1,533 @@
+    where
+        Q: BorshSerialize,
+        K: Borrow<Q>,
+    {
+        let storage_bytes = env::storage_read(&Self::lookup_key(prefix, key));
+        storage_bytes.as_deref().map(Self::deserialize_element)
+    }
+
+    /// Returns a reference to the value corresponding to the key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`BorshSerialize`] and [`ToOwned<Owned = K>`](ToOwned) on the borrowed form *must* match those for
+    /// the key type.
+    pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K>,
+    {
+        //* ToOwned bound, which forces a clone, is required to be able to keep the key in the cache
+        let entry = self
+            .cache
+            .get(k.to_owned())
+            .get_or_init(|| CacheEntry::new_cached(Self::load_element(&self.prefix, k)));
+        entry.value().as_ref()
+    }
+
+    fn get_mut_inner<Q: ?Sized>(&mut self, k: &Q) -> &mut CacheEntry<V>
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K>,
+    {
+        let prefix = &self.prefix;
+        //* ToOwned bound, which forces a clone, is required to be able to keep the key in the cache
+        let entry = self.cache.get_mut(k.to_owned());
+        entry.get_or_init(|| CacheEntry::new_cached(Self::load_element(prefix, k)));
+        let entry = entry.get_mut().unwrap_or_else(|| unreachable!());
+        entry
+    }
+
+    /// Returns a mutable reference to the value corresponding to the key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`BorshSerialize`] and [`ToOwned<Owned = K>`](ToOwned) on the borrowed form *must* match those for
+    /// the key type.
+    pub fn get_mut<Q: ?Sized>(&mut self, k: &Q) -> Option<&mut V>
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K>,
+    {
+        self.get_mut_inner(k).value_mut().as_mut()
+    }
+
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not have this key present, [`None`] is returned.
+    ///
+    /// If the map did have this key present, the value is updated, and the old
+    /// value is returned. The key is not updated, though; this matters for
+    /// types that can be `==` without being identical.
+    pub fn insert(&mut self, k: K, v: V) -> Option<V>
+    where
+        K: Clone,
+    {
+        self.get_mut_inner(&k).replace(Some(v))
+    }
+
+    /// Returns `true` if the map contains a value for the specified key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`BorshSerialize`] and [`ToOwned<Owned = K>`](ToOwned) on the borrowed form *must* match those for
+    /// the key type.
+    pub fn contains_key<Q: ?Sized>(&self, k: &Q) -> bool
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K> + Ord,
```

Comment:
Huh, that's subtle! 

I think here we have an extra requrenment that `Q: BorshSerialize` and `K: BorshSerialize` use the same serialization format. IE, this extends the usual `Borrow` contract to cover serialization format as well. 

I don't see any immediate problems here.

Reply:
yeah, it's not ideal. I couldn't think of a cleaner way to handle this, though. Maybe we can document just to make sure this assumption is clear. Assuming `K` and `Borrow<K>` have the same serialization seems in line with how `Borrow` is defined in some places, but others indicate that it only covers `Hash`, `Eq`, `Ord`